pkg/db/db.go

```diff
@@ -1,268 +1,183 @@
-
-	// github.Client
-	github.NewClient,
-	wire.Bind(new(github.Operation), new(github.Client)),
+// Operation defines the DB operations
+type Operation interface {
+	NeedsUpdate(cliVersion string, skip bool) (need bool, err error)
+	Download(ctx context.Context, dst string) (err error)
+	UpdateDownloadedAt() error
+}
 
-	// Metadata
-	afero.NewOsFs,
-	NewMetadata,
+type options struct {
+	artifact *oci.Artifact
+	clock    clock.Clock
+}
 
-	// db.Client
-	NewClient,
-	wire.Bind(new(Operation), new(Client)),
-)
+// Option is a functional option
+type Option func(*options)
 
-// Operation defines the DB operations
-type Operation interface {
-	NeedsUpdate(cliVersion string, light, skip bool) (need bool, err error)
-	Download(ctx context.Context, cacheDir string, light bool) (err error)
-	UpdateMetadata(cacheDir string) (err error)
+// WithOCIArtifact takes an OCI artifact
+func WithOCIArtifact(art *oci.Artifact) Option {
+	return func(opts *options) {
+		opts.artifact = art
+	}
 }
 
-type dbOperation interface {
-	GetMetadata() (metadata db.Metadata, err error)
-	StoreMetadata(metadata db.Metadata, dir string) (err error)
+// WithClock takes a clock
+func WithClock(clock clock.Clock) Option {
+	return func(opts *options) {
+		opts.clock = clock
+	}
 }
 
 // Client implements DB operations
 type Client struct {
-	dbc          dbOperation
-	githubClient github.Operation
-	pb           indicator.ProgressBar
-	clock        clock.Clock
-	metadata     Metadata
+	*options
+
+	cacheDir string
+	metadata metadata.Client
+	quiet    bool
 }
 
 // NewClient is the factory method for DB client
-func NewClient(dbc dbOperation, githubClient github.Operation, pb indicator.ProgressBar, clock clock.Clock, metadata Metadata) Client {
-	return Client{
-		dbc:          dbc,
-		githubClient: githubClient,
-		pb:           pb,
-		clock:        clock,
-		metadata:     metadata,
+func NewClient(cacheDir string, quiet bool, opts ...Option) *Client {
+	o := &options{
+		clock: clock.RealClock{},
 	}
-}
 
-// NeedsUpdate check is DB needs update
-func (c Client) NeedsUpdate(cliVersion string, light, skip bool) (bool, error) {
-	dbType := db.TypeFull
-	if light {
-		dbType = db.TypeLight
+	for _, opt := range opts {
+		opt(o)
+	}
+
+	return &Client{
+		options:  o,
+		cacheDir: cacheDir,
+		metadata: metadata.NewClient(cacheDir),
+		quiet:    quiet,
 	}
+}
 
-	metadata, err := c.metadata.Get()
+// NeedsUpdate check is DB needs update
+func (c *Client) NeedsUpdate(cliVersion string, skip bool) (bool, error) {
+	meta, err := c.metadata.Get()
 	if err != nil {
 		log.Logger.Debugf("There is no valid metadata file: %s", err)
 		if skip {
 			log.Logger.Error("The first run cannot skip downloading DB")
 			return false, xerrors.New("--skip-update cannot be specified on the first run")
 		}
-		metadata = db.Metadata{} // suppress a warning
+		meta = metadata.Metadata{Version: db.SchemaVersion}
 	}
 
-	if db.SchemaVersion < metadata.Version {
+	if db.SchemaVersion < meta.Version {
```

Comment:
it doesn't work well if customer reverts to previous trivy version (Like 0.21.3). It's a cache anyway, why doesn't just override it with version supported by current `trivy` binary?

Reply:
actually I think it may be useful to stop use of outdated db. But in this case missed scenario is: Version is not updated anymore but user keep using old binary. This may lead to false negatives as db will be outdated

Reply:
one of option to fix that add check like
```Now().After(meta.NextUpdate)```
meaning: "expected update has not happen"

Reply:
>one of option to fix that add check like
>Now().After(meta.NextUpdate)
>meaning: "expected update has not happen"

Yes, but we need to add the check to the older binary, right? It doesn't make sense to add the check to the new version. If you are talking about future migration like v2 to v3, you are right.

Reply:
> It's a cache anyway, why doesn't just override it with version supported by current trivy binary?

We want to encourage users to update to the latest version. We can download the v1 DB when they use an old binary, but the v1 database will be outdated and lead to false negatives as you said.

Reply:
I think the check I suggested is required (or something else). Just to notify users (eventually) that db with old schema version is not updated anymore. (if they stick to old binary version for some reasons). I can create separate issue if required.

Reply:
OK. We can do that in a dedicated PR.